Add terraform orchestration for CaaSP deployment on OpenStack cloud

mjura opened the pull request:
We would like to have CaaSP installation on OpenStack automated, terraform based can help with that.

kiall commented:
This looks like it duplicates the stuff we already have in the openstack-heat folder? But, requires tools (terraform) is installed locally vs being able to use it from both the CLI and Horizon web UIs... 

Is there any reason we need to redo this with terraform instead of adding the extra pieces (security groups) to the heat templates?

MaximilianMeister commented:
> Is there any reason we need to redo this with terraform instead of adding the extra pieces (security groups) to the heat templates?

it's easier for the devs i think as we already have terraform installed with the devenv, e.g. i don't have any openstack clients installed on my machine. maybe some documentation on the usage of `openstack-heat` would help devs, but still i like more the terraform workflow, as we're using it all the time already (devenv, caasp-kvm)

plus we could maybe use it to replace the "terraform create environment" stage?

mjura commented:
Terreform is nice alternative for Heat. It can be used to orchestrate CaaSP on different Cloud providers in easy way (libvirt, OpenStack, AWS, GCE etc) it means that it is way for CaaSP deployment standardization. Other arguments, as Max said it is nicer for development. It can be used also to scaling CaaSP deployment, after changing number of Workers you have to call only `caasp-run plan` and `caasp-run apply` and new nodes will be added without redeploying whole CaaSP cluster. I will also submit changes for Heat templates.

mjura commented:
I would like to merge it, since I use it for my deployments

kiall commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,167 @@
+variable "minions_size" {}
+variable "minions" {
+  default = 5
+}
```
Can we call these workers? both masters and workers are minions :)

mjura replied:
I've changed the name of it on `workers`

kiall commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,167 @@
+}
+
+resource "openstack_compute_instance_v2" "master" {
+  name       = "CaaSP-Master"
```
We've been updating tooling to support multiple masters, can we add a count here to allow for multiple?

mjura replied:
I've added also counter for master nodes and I'm waiting for this feature

kiall commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  security_groups = [
+    "default",
```
The default secgroup is almost universally "all everything" after people boot a few VMs, we should just rely on the groups we manage.

mjura replied:
Ok, I've added three more security groups for each node role

kiall commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  security_groups = [
+    "default",
```
The default secgroup is almost universally "all everything" after people boot a few VMs, we should just rely on the groups we manage.

kiall commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  security_groups = [
+    "default",
```
The default secgroup is almost universally "all everything" after people boot a few VMs, we should just rely on the groups we manage.

MaximilianMeister commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,167 @@
+
+output "ip_master" {
+  value = "${openstack_compute_instance_v2.master.access_ip_v4}"
+}
```
can we also add the admin ip here?

mjura replied:
I've added output also for `ip_admin`

MaximilianMeister replied:
we should also show the `workers` i think, like in `caasp-kvm`. can you still add that?

mjura replied:
I've added workers ip to output

MaximilianMeister commented:
It would also be great if we can already associate floating ip's along with the nodes, and put that into the output of terraform

MaximilianMeister commented:
@mjura i think you forgot to check in the example config file `openstack.tfvars.example`

MaximilianMeister commented:
> i think you forgot to check in the example config file openstack.tfvars.example

oh i just saw the `.openrc` support, even better thanks

mjura commented:
I switched to `openrc` file with configuration, additional options can be changed in `caasp-cluster.tf` file. I've tested it on skycastle and works fine. I haven't added floating network yet. Probably in next PR's

MaximilianMeister commented on caasp-openstack/caasp-cluster.tf:
```diff
@@ -0,0 +1,275 @@
+variable "user_name" {}
+variable "password" {}
+variable "image_name" {
+  default = "CaaSP-1.0-Build30.25-GMC"
```
i think we need some tfvars file after all, if only because of the image name

MaximilianMeister commented:
MINOR: and @mjura could you rename `caasp-run` to `caasp-openstack` ? that would be the same schema like `caasp-kvm`

mjura commented:
@MaximilianMeister: I've addressed all your comments. Floating IP's I will add later, I would like to have this initial change merge first. I've tested it and works fine. I use it all over the time for testing CaaSP

kiall commented on caasp-openstack/README.md:
````diff
@@ -0,0 +1,37 @@
+
+Deploy CaaSP on OpenStack
+```
+./caasp-run apply
````
Minor: these should now be `./caasp-openstack ...`

mjura replied:
Fixed

kiall commented on caasp-openstack/openstack.tfvars:
```diff
@@ -0,0 +1,7 @@
+master_size = "m1.large"
+masters = 1
+worker_size = "m1.large"
+workers = 3
```
minor: setting this to 2 gives a valid, but smaller, cluster :)

mjura replied:
Ok, change this default to 2

mjura commented:
Ok, I've fixed minor comments. Thank you for reviewing
